BungeeCord-Patches/0061-Rework-information-forwarding.patch

```diff
@@ -0,0 +1,371 @@
+\ No newline at end of file
+diff --git a/api/src/main/java/net/md_5/bungee/Util.java b/api/src/main/java/net/md_5/bungee/Util.java
+index 77eb64a1..8ffb37e7 100644
+--- a/api/src/main/java/net/md_5/bungee/Util.java
++++ b/api/src/main/java/net/md_5/bungee/Util.java
+@@ -8,6 +8,8 @@ import java.net.InetSocketAddress;
+ import java.net.SocketAddress;
+ import java.net.URI;
+ import java.net.URISyntaxException;
++import java.security.SecureRandom;
++import java.util.Random;
+ import java.util.UUID;
+ 
+ import io.github.waterfallmc.waterfall.utils.Hex;
+@@ -106,4 +108,22 @@ public class Util
+     {
+         return new UUID( UnsignedLongs.parseUnsignedLong( uuid.substring( 0, 16 ), 16 ), UnsignedLongs.parseUnsignedLong( uuid.substring( 16 ), 16 ) );
+     }
++
++    // Waterfall start: Forwarding rework
++    /**
++     * Generates an alphanumeric A-Z,a-z,0-9 byte-sequence.
++     *
++     * @param len the length of the sequence
++     * @return a UTF/ASCII compatible alphanumeric byte-sequence
++     */
++    public static byte[] randomAlphanumericSequence(int len){
++        Random random = new SecureRandom();
++        byte[] ret = new byte[len];
++        for(int i = 0; i < len; i++){
++            int seq = random.nextInt(62);
++            ret[i] = (byte) (seq < 10 ? seq + 48 : seq < 36 ? seq + 55 : seq + 61) ;
++        }
++        return ret;
++    }
++    // Waterfall end: Forwarding rework
+ }
+diff --git a/api/src/main/java/net/md_5/bungee/api/ProxyConfig.java b/api/src/main/java/net/md_5/bungee/api/ProxyConfig.java
+index d69463f0..59f61e47 100644
+--- a/api/src/main/java/net/md_5/bungee/api/ProxyConfig.java
++++ b/api/src/main/java/net/md_5/bungee/api/ProxyConfig.java
+@@ -2,6 +2,8 @@ package net.md_5.bungee.api;
+ 
+ import java.util.Collection;
+ import java.util.Map;
++
++import io.github.waterfallmc.waterfall.forwarding.ForwardingMode;
+ import net.md_5.bungee.api.config.ListenerInfo;
+ import net.md_5.bungee.api.config.ServerInfo;
+ 
+@@ -261,4 +263,10 @@ public interface ProxyConfig
+      * @return {@code true} if tablist rewriting is disabled, {@code false} otherwise
+      */
+     boolean isDisableTabListRewrite();
++
++    /**
++     * Represents the forwarding mode as configured.
++     * @return the mode set in the config
++     */
++    ForwardingMode getForwardingMode();
+ }
+diff --git a/api/src/main/java/net/md_5/bungee/api/plugin/PluginManager.java b/api/src/main/java/net/md_5/bungee/api/plugin/PluginManager.java
+index 90031156..f1aad373 100644
+--- a/api/src/main/java/net/md_5/bungee/api/plugin/PluginManager.java
++++ b/api/src/main/java/net/md_5/bungee/api/plugin/PluginManager.java
+@@ -439,6 +439,14 @@ public final class PluginManager
+                         Preconditions.checkNotNull( desc.getName(), "Plugin from %s has no name", file );
+                         Preconditions.checkNotNull( desc.getMain(), "Plugin from %s has no main", file );
+ 
++                        // Waterfall start: Forwarding rework
```

Comment:
I err on this, no idea if others have any other thoughts on this, I mean, it makes sense but I don't want this to become a norm in any form of the sense